File: src/morie/fn/ipfsfa.py
# morie.fn -- wave2 slice w2_02 (rootcoder007/morie)
"""Barrier interior-point solver for smooth inequality-constrained NLPs.

Waechter and Biegler (2006), "On the implementation of an
interior-point filter line-search algorithm for large-scale nonlinear
programming", Mathematical Programming 106(1):25-57,
doi:10.1007/s10107-004-0559-y -- the IPOPT paper.  The core of that
algorithm is the barrier subproblem

    minimise  f(x) - mu sum_i log(-g_i(x)),   g_i(x) <= 0,

solved approximately for a decreasing sequence of mu with a Newton
step and a backtracking line search that keeps the iterate strictly
feasible.  Derivatives are taken by central differences with a fixed
step, so no symbolic gradient is required of the caller and both
language arms follow the same trajectory.
"""

from __future__ import annotations

import math

from . import _array_core as np  # noqa: F401
from . import _s03core as core

from ._richresult import RichResult

__all__ = ["ipopt_solver"]

_H = 1e-5


def _phi(f, cons, x, mu):
    v = float(f(x))
    for g in cons:
        gv = float(g(x))
        if gv >= 0:
            return float("inf")
        v -= mu * math.log(-gv)
    return v


def ipopt_solver(f, constraints, x0, mu0=1.0, outer=8, inner=30):
    """Minimise f subject to g_i(x) <= 0 by a decreasing-mu barrier."""
    x = core.vec(x0)
    n = len(x)
    if n == 0:
        raise ValueError("ipopt_solver: x0 is empty")
    if not callable(f):
        raise ValueError("ipopt_solver: f must be callable")
    cons = list(constraints)
    for g in cons:
        if not callable(g):
            raise ValueError("ipopt_solver: every constraint must be callable")
        if float(g(x)) >= 0:
            raise ValueError("ipopt_solver: x0 must be strictly feasible")
    mu = float(mu0)
    if mu <= 0:
        raise ValueError("ipopt_solver: mu0 must be positive")
    for _ in range(int(outer)):
        for _ in range(int(inner)):
            base = _phi(f, cons, x, mu)
            g = [0.0] * n
            for j in range(n):
                xp = list(x)
                xm = list(x)
                xp[j] += _H
                xm[j] -= _H
                g[j] = (_phi(f, cons, xp, mu) - _phi(f, cons, xm, mu)) / (2.0 * _H)
            H = [[0.0] * n for _ in range(n)]
            for j in range(n):
                xp = list(x)
                xm = list(x)
                xp[j] += _H
                xm[j] -= _H
                H[j][j] = (_phi(f, cons, xp, mu) - 2.0 * base + _phi(f, cons, xm, mu)) / (_H * _H)
                for k in range(j + 1, n):
                    xpp = list(x)
                    xpm = list(x)
                    xmp = list(x)
                    xmm = list(x)
                    xpp[j] += _H
                    xpp[k] += _H
                    xpm[j] += _H
                    xpm[k] -= _H
                    xmp[j] -= _H
                    xmp[k] += _H
                    xmm[j] -= _H
                    xmm[k] -= _H
                    v = (_phi(f, cons, xpp, mu) - _phi(f, cons, xpm, mu) - _phi(f, cons, xmp, mu) + _phi(f, cons, xmm, mu)) / (4.0 * _H * _H)
                    H[j][k] = v
                    H[k][j] = v
            for j in range(n):
                H[j][j] += 1e-8
            try:
                step = core.cholsolve(H, [-v for v in g])
            except Exception:
                step = [-v for v in g]
            a = 1.0
            ok = False
            for _ in range(60):
                xn = [x[j] + a * step[j] for j in range(n)]
                if _phi(f, cons, xn, mu) < base:
                    x = xn
                    ok = True
                    break
                a *= 0.5
            if not ok:
                break
        mu *= 0.2
    viol = max((float(g(x)) for g in cons), default=-1.0)
    return RichResult(
        title="Barrier interior-point solver",
        summary_lines=[("variables", n), ("constraints", len(cons))],
        payload={
            "estimate": float(f(x)),
            "x": x,
            "objective": float(f(x)),
            "max_violation": viol,
            "mu_final": mu,
            "n": n,
            "method": "decreasing-mu log-barrier with Newton steps, Waechter & Biegler (2006)",
        },
    )


def cheatsheet():
    return "ipfsfa: barrier interior-point NLP solver"


# compact alias per ledger/NAMING.md
ipoptsolver = ipopt_solver

File: r-package/morie/R/Ipfsfa.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Barrier interior-point solver for smooth inequality-constrained problems
#'
#' The core of the IPOPT algorithm: a barrier subproblem solved
#' approximately for a decreasing sequence of mu, with a Newton step and
#' a backtracking line search that keeps the iterate strictly feasible.
#' Derivatives are central differences with a fixed step, so no
#' symbolic gradient is required and both language arms follow the same
#' trajectory.
#'
#' Formula: minimise f(x) - mu sum_i log(-g_i(x)) with g_i(x) <= 0.
#'
#' @param f Objective function of a numeric vector.
#' @param constraints List of functions g_i with g_i(x) <= 0 feasible.
#' @param x0 Strictly feasible starting point.
#' @param mu0 Initial barrier parameter.
#' @param outer Number of mu reductions.
#' @param inner Newton steps per barrier subproblem.
#' @return List with \code{estimate}, \code{x}, \code{objective},
#'   \code{max_violation}, \code{mu_final}, \code{n}, \code{method}.
#' @references Waechter and Biegler (2006), On the implementation of an
#'   interior-point filter line-search algorithm for large-scale
#'   nonlinear programming, Mathematical Programming 106(1):25-57.
#'   \doi{10.1007/s10107-004-0559-y}
#' @export
Ipfsfa <- function(f, constraints, x0, mu0 = 1, outer = 8, inner = 30) {
  x <- .s03vec(x0)
  n <- length(x)
  if (n == 0L) stop("ipopt_solver: x0 is empty")
  if (!is.function(f)) stop("ipopt_solver: f must be callable")
  cons <- as.list(constraints)
  for (g in cons) {
    if (!is.function(g)) stop("ipopt_solver: every constraint must be callable")
    if (as.numeric(g(x)) >= 0) stop("ipopt_solver: x0 must be strictly feasible")
  }
  mu <- as.numeric(mu0)
  if (mu <= 0) stop("ipopt_solver: mu0 must be positive")
  hh <- 1e-5
  phi <- function(x, mu) {
    v <- as.numeric(f(x))
    for (g in cons) {
      gv <- as.numeric(g(x))
      if (gv >= 0) return(Inf)
      v <- v - mu * log(-gv)
    }
    v
  }
  for (o in seq_len(as.integer(outer))) {
    for (k in seq_len(as.integer(inner))) {
      base <- phi(x, mu)
      g <- numeric(n)
      for (j in seq_len(n)) {
        xp <- x; xm <- x; xp[j] <- xp[j] + hh; xm[j] <- xm[j] - hh
        g[j] <- (phi(xp, mu) - phi(xm, mu)) / (2 * hh)
      }
      H <- matrix(0, n, n)
      for (j in seq_len(n)) {
        xp <- x; xm <- x; xp[j] <- xp[j] + hh; xm[j] <- xm[j] - hh
        H[j, j] <- (phi(xp, mu) - 2 * base + phi(xm, mu)) / (hh * hh)
        if (j < n) for (k2 in seq(j + 1L, n)) {
          xpp <- x; xpm <- x; xmp <- x; xmm <- x
          xpp[j] <- xpp[j] + hh; xpp[k2] <- xpp[k2] + hh
          xpm[j] <- xpm[j] + hh; xpm[k2] <- xpm[k2] - hh
          xmp[j] <- xmp[j] - hh; xmp[k2] <- xmp[k2] + hh
          xmm[j] <- xmm[j] - hh; xmm[k2] <- xmm[k2] - hh
          v <- (phi(xpp, mu) - phi(xpm, mu) - phi(xmp, mu) + phi(xmm, mu)) / (4 * hh * hh)
          H[j, k2] <- v; H[k2, j] <- v
        }
      }
      diag(H) <- diag(H) + 1e-8
      step <- tryCatch(.s03cholsolve(H, -g), error = function(e) -g)
      a <- 1; okstep <- FALSE
      for (i in seq_len(60)) {
        xn <- x + a * step
        if (phi(xn, mu) < base) { x <- xn; okstep <- TRUE; break }
        a <- a / 2
      }
      if (!okstep) break
    }
    mu <- mu * 0.2
  }
  viol <- if (length(cons)) max(vapply(cons, function(g) as.numeric(g(x)), 0)) else -1
  .t1_result(estimate = as.numeric(f(x)), x = x, objective = as.numeric(f(x)),
             max_violation = viol, mu_final = mu, n = n,
             method = "decreasing-mu log-barrier with Newton steps, Waechter & Biegler (2006)")
}
